File: ThreeJS/src/game/pinball-knight/run/grade.ts
/**
 * The floor's score: FLOW and the end-of-floor GRADE.
 *
 * Extracted verbatim from core.ts. Pure reads off `state` plus the tuning
 * table — no scene, no THREE, no lifecycle — which is what makes it the one
 * piece of the descent path that is straightforwardly testable.
 */
import { state } from "../state";
import {
  GRADE_FLOW_FULL,
  GRADE_FLOW_OK,
  GRADE_KILLS_FULL,
  GRADE_KILLS_OK,
  GRADE_COMBO_FULL,
  GRADE_COMBO_OK,
  GRADE_GOLD,
} from "../constants";

export function floorFlow(): number {
  if (state.levelFlowT <= 0) return 0;
  return state.levelFlowSum / state.levelFlowT;
}

/**
 * Grade the floor being left: FLOW (how much speed you actually carried),
 * carnage (share of the horde killed) and style (best bounce combo), two marks
 * each → S/A/B/C/D and a gold bonus. The "play it again, but cooler" hook.
 *
 * Both the pace and style axes were rebuilt in the de-clone wave. Pace was raw
 * wall-clock, so walking a floor briskly scored the same as riding it, and
 * style capped at combo 8 — the exact point where the combo curve starts being
 * interesting. Both now measure the thing the game is actually about.
 */
export function gradeFloor(): { grade: string; gold: number } {
  const kills = state.kills - state.levelStartKills;
  const share = kills / Math.max(1, state.levelHordeSize);
  const flow = floorFlow();
  let pts = 0;
  pts += flow >= GRADE_FLOW_FULL ? 2 : flow >= GRADE_FLOW_OK ? 1 : 0;
  pts += share >= GRADE_KILLS_FULL ? 2 : share >= GRADE_KILLS_OK ? 1 : 0;
  pts += state.levelBestCombo >= GRADE_COMBO_FULL ? 2 : state.levelBestCombo >= GRADE_COMBO_OK ? 1 : 0;
  const grade = pts >= 6 ? "S" : pts >= 5 ? "A" : pts >= 3 ? "B" : pts >= 2 ? "C" : "D";
  return { grade, gold: GRADE_GOLD[grade] ?? 0 };
}

File: Rust/crates/pk-core/src/run/grade.rs
//! End-of-Floor Performance Grading Rubric — FLOW velocity averaging and 3-axis combat scoring.
//!
//! PORTS: `run/grade.ts`

pub const GRADE_FLOW_FULL: f32 = 6.5;
pub const GRADE_FLOW_OK: f32 = 4.0;
pub const GRADE_KILLS_FULL: f32 = 0.90;
pub const GRADE_KILLS_OK: f32 = 0.60;
pub const GRADE_COMBO_FULL: u32 = 12;
pub const GRADE_COMBO_OK: u32 = 6;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FloorGradeResult {
    pub grade: String,
    pub points: u32,
    pub gold: u32,
}

/// Computes average speed carried through the floor (flow velocity).
pub fn compute_floor_flow(level_flow_sum: f32, level_flow_t: f32) -> f32 {
    if level_flow_t <= 0.0 {
        0.0
    } else {
        level_flow_sum / level_flow_t
    }
}

/// Grades the completed floor across FLOW, Carnage (kill share), and Style (best bounce combo).
pub fn grade_floor(kills: u32, horde_size: u32, flow: f32, best_combo: u32) -> FloorGradeResult {
    let share = kills as f32 / horde_size.max(1) as f32;
    let mut pts = 0;

    // Axis 1: Flow (pace & speed maintenance)
    if flow >= GRADE_FLOW_FULL {
        pts += 2;
    } else if flow >= GRADE_FLOW_OK {
        pts += 1;
    }

    // Axis 2: Carnage (share of horde eliminated)
    if share >= GRADE_KILLS_FULL {
        pts += 2;
    } else if share >= GRADE_KILLS_OK {
        pts += 1;
    }

    // Axis 3: Style (longest wall/actor bounce combo)
    if best_combo >= GRADE_COMBO_FULL {
        pts += 2;
    } else if best_combo >= GRADE_COMBO_OK {
        pts += 1;
    }

    let (grade_str, gold) = match pts {
        6..=u32::MAX => ("S", 100),
        5 => ("A", 60),
        3 | 4 => ("B", 35),
        2 => ("C", 15),
        _ => ("D", 0),
    };

    FloorGradeResult {
        grade: grade_str.to_string(),
        points: pts,
        gold,
    }
}
